File: readersWritersJava/src/main/java/Lightswitch.java
import java.util.concurrent.Semaphore;

public class Lightswitch {

    private int counter = 0;
    private Semaphore mutex;

    public Lightswitch() {
        this.mutex = new Semaphore(1);
    }

    public void lock(Semaphore semaphore) throws InterruptedException {
        try {
            mutex.acquire();

            counter++;
            if (counter == 1) {
                semaphore.acquire();
            }
        } finally {
            mutex.release();
        }
    }

    public void unlock(Semaphore semaphore) throws InterruptedException {
        try {
            mutex.acquire();

            counter--;
            if (counter == 0) {
                semaphore.release();
            }
        } finally {
            mutex.release();
        }
    }
}

File: readersWritersGo/lightswitch.go
package main

// Implements the Lightswitch pattern

import "sync"

// Lightswitch definition
type Lightswitch struct {
	counter int
	mutex   sync.Mutex
}

// Lock locks the semaphore if its the first thread to call this function
func (l *Lightswitch) Lock(semaphore *sync.Mutex) {
	l.mutex.Lock()
	defer l.mutex.Unlock()

	l.counter++
	if l.counter == 1 {
		semaphore.Lock()
	}
}

// Unlock unlocks the semaphore when the last thread calls this function
func (l *Lightswitch) Unlock(semaphore *sync.Mutex) {
	l.mutex.Lock()
	defer l.mutex.Unlock()

	l.counter--
	if l.counter == 0 {
		semaphore.Unlock()
	}
}
